lib/tasks/grade.rake

```diff
@@ -92,6 +97,40 @@ namespace :grade do
 
 end
 
+def sync_specs_with_source
+  base_url = "https://api.github.com/repos/"
+  reponame = Dir.pwd.split("/").last
+  url_for_spec_folder = base_url + "appdev-projects/#{reponame}/contents/spec"
+  spec_folder_json = URI.open(url_for_spec_folder).read
+  spec_folder_contents = JSON.parse(spec_folder_json)
```

Comment:
The API docs for `contents` shows a SHA in the example response. This might be because the example is for a file, but will the API endpoint for a folder show a SHA also? If so then perhaps we can speed things up by comparing the SHA of the local folder with that of the remote folder.

(I'm not set up right now to test what an API response for a folder looks like.)

Reply:
Yeah, the spec folder does have a SHA (it's what I tried first), but I didn't know how to get the SHA of the folder locally to compare.

Reply:
I think this ought to work (shelling out with backticks)?

```
`git ls-files -s spec/ | git hash-object --stdin`
```

Reply:
@jelaniwoods This might provide a hint:

https://stackoverflow.com/questions/55678890/compute-github-api-file-sha

Reply:
I was able to sort-of find the correct SHA of the sub-folder locally, with `git ls-tree HEAD spec/`

![image](https://user-images.githubusercontent.com/17581658/194366460-1b455eb9-e45c-46ef-a53c-4c323e818816.png)

which _does_ match the SHA of the folder on GitHub

![image](https://user-images.githubusercontent.com/17581658/194369977-8834cdaf-dcb6-4339-97c4-47261a040180.png)

However,  this SHA doesn't seem to update when there's uncommitted changes (because it's looking at HEAD?) which would mean we'd still need to check the status/potentially each file name.

> The SHA1 is computed based on file contents, so it should include any uncommitted changes as well. 

This definitely seems possible to do, I'm just not sure how to do it yet. It's tricky for folders b/c it's made with a combination of SHA1 of it's contents. 

> If you're interested in creating the git tree object manually from its contents, it's important to note that the [Entries having references to other trees and blobs] needs to be sorted alphabetically by [file/folder name] (aka path)
- [Investigating this](https://stackoverflow.com/a/21599232/10481804)

Hopefully, there is a built-in git way to do this.

Reply:
> Hopefully, there is a built-in git way to do this.

It seems like there would be a way to do it on the current state of the files rather than just the committed state.

A brute force solution might be to:

 - checkout a temp branch
 - commit
 - calculate the SHA
 - reset
 - checkout the old branch
 - delete the temp branch

Not sure if that would be any faster than the current method, though.

Reply:
On second thought:

> Git normally creates a tree by taking the state of your staging area or index and writing a series of tree objects from it. So, to create a tree object, you first have to set up an index by staging some files.
- [Tree Objects - Git Docs](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#Tree-Objects)

This makes it sound like Git _can't_ create a tree using unstaged changes, even if we want it to and we'd need to either do some Git trickery to temporarily stage unsaved changes, or write the code ourselves (which doesn't seem like it'd be much less performant than the current check for unsaved changes).

Reply:
We don't want to allow any changes in the folder anyway, right? How about discarding any changes (staged or unstaged) and then checking the Sha?

Reply:
I was able to get it to work with the steps you mentioned previously:

> A brute force solution might be to:
> 
>     checkout a temp branch
>     commit
>     calculate the SHA
>     reset
>     checkout the old branch
>     delete the temp branch

Just cleaning up some debugging variables and will re-request review.